Add types for representing fee rates

halseth opened the pull request:
This PR adds types for representing the different fee rates used in the code base. This helps in ensuring that the correct fee rates are applied, and simplifies the conversion between them.

halseth commented on test_utils.go:
```diff
@@ -138,18 +138,18 @@ func createTestPeer(notifier chainntnfs.ChainNotifier,
 		LocalBalance:  lnwire.NewMSatFromSatoshis(channelBal),
 		RemoteBalance: lnwire.NewMSatFromSatoshis(channelBal),
-		FeePerKw:      feePerKw,
+		FeePerKw:      btcutil.Amount(feePerKw),
```
Yeah, I wanted to do that, but the compiler throws the error "cyclic imports not allowed". I think it's because `lnwallet` imports `channeldb`, and if I make `FeePerKw` this type, then `channeldb` will import `lnwallet`, causing the cycle. 

Wondering if I should put the types / fee-estimator in a new package? Or if there's a better way to resolve this? :D

Roasbeef replied:
Yeh the route to resolve this would be to extract all the fee based stuff into a distinct package. Since the import cycle issue would only be encountered in this test file, and it isn't a blocker for this PR, I think we can postpone this movement until later. Generally the `lnwallet` package needs to be broken up, so when we lay out a plan for doing so, we should also move out all the fee estimation related stuff.

halseth replied:
The same cast is done in `lnwallet/channel.go`, so not only test files are affected. Will add a TODO for this there.

halseth commented on lnrpc/rpc.proto:
```diff
@@ -873,11 +873,11 @@ message OpenChannelRequest {
-    /// A manual fee rate set in sat/byte that should be used when crafting the closure transaction.
-    int64 sat_per_byte = 7;
+    /// A manual fee rate set in sat/byte that should be used when crafting the funding transaction.
+    int64 sat_per_byte = 7 [json_name = "sat_per_byte"];
```
Woops, this commit was not meant to be part of this PR :P  Reverted.

halseth commented on lnwallet/fee_estimator.go:
```diff
@@ -8,19 +8,68 @@ import (
+	return SatPerKWeight(s * 1000)
+}
+
+// TODO: call Kw our Kwu?
```
Using `Kw` would make it easier to distinguish from `SatPerWeight` though.

halseth replied:
Keeping `SatPerKWeight` then :)

halseth commented on lnwallet/fee_estimator.go:
```diff
@@ -8,19 +8,68 @@ import (
+
+	// If the fee rate is non-zero, then we should
+	// always return at least 1.
+	if s > 0 && f < 1 {
```
TODO: I realized this can be solved by just always rounding up. Will look into if that takes care of all these cases.

cfromknecht commented on lnwallet/fee_estimator.go:
```diff
@@ -8,19 +8,57 @@ import (
+func (s SatPerKWeight) FeePerWeight() SatPerWeight {
+	// We always round up when converting, ensuring the fee rate
+	// doesn't drop unexpectedly.
+	return SatPerWeight((s + 999) / 1000)
```
Even better :)

Roasbeef commented:
@halseth once you rebase I'll get this merged! (and will also merge the related `btcwallet` PR).

halseth commented:
I rewrote this PR a little bit, to simplify the actual `feeRate->fee` calculation, and fix some test failures resulting from the (now) more accurate fee conversions.

Number of fee rate types was reduced to two: 1) `FeePerVSize` and 2) `FeePerKWeight`, as those are the two we actually need for 1) our own fee estimation, 2) LN protocol fee rates. I added methods to these for calculating the resulting fee given a `vsize` / `weight`, and removed the lossy conversions between them, making you less likely to lose precision (I did this mistake myself, causing me a lot of pain in one of the failing tests lol).

Also rebased on master.

wilmerpaulino commented on lnwallet/wallet.go:
```diff
@@ -97,11 +97,11 @@ type initFundingReserveMsg struct {
-	// fundingFeePerWeight is the fee rate in satoshis per eight unit to
-	// use for the initial funding transaction.
-	fundingFeePerWeight btcutil.Amount
+	// fundingFeePerVSize is the fee rate in sat/vbyt to use for the initial
```
Nit: typo missing "e" in `sat/vbyte`

Roasbeef commented on channeldb/channel.go:
```diff
@@ -249,6 +249,10 @@ type ChannelCommitment struct {
+	//
+	// TODO(halseth): make this SatPerKWeight. Cannot be done atm because
+	// this will cause the import cycle lnwallet<->channeldb. Fee
+	// estimation stuff should be in its own package.
```
👍

Roasbeef commented on htlcswitch/link.go:
```diff
@@ -387,16 +386,16 @@ func (l *channelLink) EligibleToForward() bool {
 	// We'll first query for the sat/weight recommended to be confirmed
 	// within 3blocks.
-	feePerWeight, err := l.cfg.FeeEstimator.EstimateFeePerWeight(3)
+	feePerVSize, err := l.cfg.FeeEstimator.EstimateFeePerVSize(3)
```
Also a change to fix the typo above: 3blocks -> 3 blocks

Roasbeef commented on lnwallet/fee_estimator.go:
```diff
@@ -8,19 +8,38 @@ import (
+
+// FeeForVSize calculates the fee resulting from this fee rate and
+// the given vsize.
+func (s SatPerVByte) FeeForVSize(vsize int64) btcutil.Amount {
```
Dig the simplicity of these new types 😎

halseth commented on lnwallet/channel.go:
```diff
@@ -1626,14 +1626,14 @@ func (lc *LightningChannel) restoreStateLogs(
 		// on.
 		if feeUpdate, ok := logUpdate.UpdateMsg.(*lnwire.UpdateFee); ok {
-			newFeeRate := btcutil.Amount(feeUpdate.FeePerKw)
+			newFeeRate := SatPerKWeight(feeUpdate.FeePerKw)
```
Yeah, unfortunately. Added `todo` for changing it later 👍

halseth commented on lnwallet/channel_test.go:
```diff
@@ -2100,7 +2100,7 @@ func TestUpdateFeeSenderCommits(t *testing.T) {
 	}
 
-	if bobChannel.channelState.LocalCommitment.FeePerKw != fee {
+	if bobChannel.channelState.LocalCommitment.FeePerKw != btcutil.Amount(fee) {
```
Good catch! I changed which var I casted, probably missed a few cases :)

halseth commented on htlcswitch/link_test.go:
```diff
@@ -2754,9 +2754,13 @@ func TestChannelLinkUpdateCommitFee(t *testing.T) {
 	startingFeeRate := channels.aliceToBob.CommitFeeRate()
 
+	// Convert starting fee rate to sat/vbyte.
+	startingFeeRateSatPerVByte := lnwallet.SatPerVByte(
```
Added comment why we do this here.

halseth commented on fundingmanager.go:
```diff
@@ -862,7 +862,7 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 	chainHash := chainhash.Hash(msg.ChainHash)
 	reservation, err := f.cfg.Wallet.InitChannelReservation(amt, 0,
-		msg.PushAmount, btcutil.Amount(msg.FeePerKiloWeight), 0,
+		msg.PushAmount, lnwallet.SatPerKWeight(msg.FeePerKiloWeight), 0,
```
Same, import cycle. Added `todo` :)
